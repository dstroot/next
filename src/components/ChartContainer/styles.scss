// make the slider handle larger
.rc-slider-handle {
  width: 20px;
  height: 20px;
  margin-left: -10px; // half width
  margin-top: -8px; // half height allowing for 4px bar width
}
