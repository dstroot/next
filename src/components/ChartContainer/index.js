import React, { useState, useEffect } from 'react';

// slider
import Slider, { createSliderWithTooltip } from 'rc-slider';
import 'rc-slider/assets/index.css';

// chart
import Chart from '../Chart';

// images and styles

import './styles.scss';

const SliderWithTooltip = createSliderWithTooltip(Slider);

// This function calculates future value.  This function matches the
// MSFT Excel fv calculation.
function futureValue(rate, nper, pmt, pv, type) {
  let pow = Math.pow(1 + rate, nper);
  let fv = 0;

  if (rate) {
    fv = (pmt * (1 + rate * type) * (1 - pow)) / rate - pv * pow;
  } else {
    fv = -1 * (pv + pmt * nper);
  }

  return fv.toFixed(2);
}

const ChartContainer = () => {
  // calculate a logarithmic slider position from a value
  const logPosInitial = value => {
    // position will be between 0 and 100
    var minp = 0;
    var maxp = 99;

    // The result should be between 500 and 1,000,000
    var minv = Math.log(500);
    var maxv = Math.log(1000000);

    // calculate adjustment factor
    var scale = (maxv - minv) / (maxp - minp);

    return (Math.log(value) - minv) / scale + minp;
  };

  // need to make the slider logarithmic
  // https://stackoverflow.com/questions/846221/logarithmic-slider
  const logSlideInitial = position => {
    // position will be between 0 and 100
    var minp = 0;
    var maxp = 99;

    // The result should be between 500 and 1,000,000
    var minv = Math.log(500);
    var maxv = Math.log(1000000);

    // calculate adjustment factor
    var scale = (maxv - minv) / (maxp - minp);

    // calculate result
    let result = Math.exp(minv + scale * (position - minp));

    // round to nearest - but for a logarithmic scale!
    if (result < 1000) {
      result = Math.ceil(result / 500) * 500;
    }
    if (result >= 1000 && result < 10000) {
      result = Math.ceil(result / 1000) * 1000;
    }
    if (result >= 10000 && result < 100000) {
      result = Math.ceil(result / 10000) * 10000;
    }
    if (result >= 100000 && result < 999999) {
      result = Math.ceil(result / 100000) * 100000;
    }

    return result;
  };

  // calculate a logarithmic slider position from a value
  const logPosMonthly = value => {
    // position will be between 0 and 100
    var minp = 0;
    var maxp = 99;

    // The result should be between 25 and 25,000
    var minv = Math.log(25);
    var maxv = Math.log(25000);

    // calculate adjustment factor
    var scale = (maxv - minv) / (maxp - minp);

    return (Math.log(value) - minv) / scale + minp;
  };

  // need to make the slider logarithmic
  // https://stackoverflow.com/questions/846221/logarithmic-slider
  const logSlideMonthy = position => {
    // position will be between 0 and 100
    var minp = 0;
    var maxp = 99;

    // The result should be between 25 and 25,000
    var minv = Math.log(25);
    var maxv = Math.log(25000);

    // calculate adjustment factor
    var scale = (maxv - minv) / (maxp - minp);

    // calculate result
    let result = Math.exp(minv + scale * (position - minp));

    // round to nearest - but for a logarithmic scale!
    if (result < 50) {
      result = Math.ceil(result / 5) * 5;
    }
    if (result >= 50 && result < 100) {
      result = Math.ceil(result / 10) * 10;
    }
    if (result >= 100 && result < 1000) {
      result = Math.ceil(result / 100) * 100;
    }
    if (result >= 1000 && result < 10000) {
      result = Math.ceil(result / 1000) * 1000;
    }
    if (result >= 10000 && result < 24999) {
      result = Math.ceil(result / 5000) * 5000;
    }

    return result;
  };

  const [initial, setInitial] = useState(50000);
  const [monthly, setMonthly] = useState(1000);
  const [risk, setRisk] = useState(4);
  const [riskDesc, setRiskDesc] = useState('Growth');
  const [data, setData] = useState([]);

  useEffect(() => {
    setData(createGraphData(initial, monthly, risk));
  }, [initial, monthly, risk]);

  // currency formatter
  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  });

  // This function creates graph data.  It takes inputs from the sliders.
  function createGraphData(initialDeposit, monthlyDeposit, riskLevel) {
    const type = 0; // When payments are due: 0 = end of period, 1 = beginning of period.
    const data = []; // Array of data we will return

    // table to define rates to use by risk level
    var rates = [
      { risk: 0, savings: 0.028, investment: 0.03, annuity: 0.03225 },
      { risk: 1, savings: 0.028, investment: 0.0325, annuity: 0.03475 },
      { risk: 2, savings: 0.028, investment: 0.035, annuity: 0.03725 },
      { risk: 3, savings: 0.028, investment: 0.0375, annuity: 0.03975 },
      { risk: 4, savings: 0.028, investment: 0.04, annuity: 0.04225 },
      { risk: 5, savings: 0.028, investment: 0.0425, annuity: 0.04475 },
      { risk: 6, savings: 0.028, investment: 0.045, annuity: 0.04725 },
      { risk: 7, savings: 0.028, investment: 0.0475, annuity: 0.04975 },
      { risk: 8, savings: 0.028, investment: 0.05, annuity: 0.05225 },
      { risk: 9, savings: 0.028, investment: 0.0525, annuity: 0.05475 },
    ];

    // we need to turn "risk level" into return expectations. Risk level can vary between 0 and 9.
    // Risk 0 might be equal to treasuries (e.g. about 2.5-3%) and each tick up might add .25% or so.
    // For the savings rate we might assume a constant rate over time regardless of risk.  CD rates
    // are about 2.8% right now.
    var filteredRates = rates.filter(function(rate) {
      return rate.risk === riskLevel;
    });

    // create 30 years of data
    for (let year = 0; year <= 30; year++) {
      // calculate the fv for each year
      let savings = -futureValue(
        filteredRates[0].savings / 12, // convert annual rate to monthky rate
        year * 12, // convert years to months
        monthlyDeposit,
        initialDeposit,
        type
      );

      let investment = -futureValue(
        filteredRates[0].investment / 12,
        year * 12,
        monthlyDeposit,
        initialDeposit,
        type
      );

      let annuity = -futureValue(
        filteredRates[0].annuity / 12,
        year * 12,
        monthlyDeposit,
        initialDeposit,
        type
      );

      investment = investment - savings;
      annuity = annuity - investment - savings;

      data.push({ year, savings, investment, annuity });
    }

    return data;
  }

  const changeInitial = value => {
    setInitial(logSlideInitial(value));
  };

  const changeMonthly = value => {
    setMonthly(logSlideMonthy(value));
  };

  const changeRisk = value => {
    setRisk(value);

    // switch to set risk description
    switch (value) {
      case 0:
      case 1:
      case 2:
        setRiskDesc('Conservative');
        break;

      case 3:
      case 4:
      case 5:
        setRiskDesc('Balanced');
        break;

      case 6:
      case 7:
      case 8:
        setRiskDesc('Growth');
        break;

      case 9:
        setRiskDesc('Aggressive');
        break;

      default:
        setRiskDesc('Error');
        break;
    }
  };

  return (
    <div className="container mt-5 mb-4">
      <div className="row">
        <div className="col-md-12" />
        <div className="col-md-12 text-center">
          <h1>Try Yourself</h1>
          <p>
            Use the slider to see how monthly income can change based on the
            monthly contribution amount and income start date.
          </p>
          <div className="mt-4">
            <div className="d-flex justify-content-between">
              <div>Initial Deposit</div>
              <div>
                <h4 className="text-primary">
                  {formatter.format(initial).slice(0, -3)}
                </h4>
              </div>
            </div>
            <Slider
              min={0}
              max={99}
              defaultValue={logPosInitial(50000)}
              onChange={changeInitial}
            />
          </div>
          <div className="mt-4">
            <div className="d-flex justify-content-between">
              <div>Monthly Deposit</div>
              <div>
                <h4 className="text-primary">
                  {formatter.format(monthly).slice(0, -3)}
                </h4>
              </div>
            </div>
            <Slider
              min={0}
              max={99}
              defaultValue={logPosMonthly(1000)}
              onChange={changeMonthly}
            />
          </div>
          <div className="mt-4">
            <div className="d-flex justify-content-between">
              <div>Risk Level</div>
              <div>
                <h4 className="text-primary">{riskDesc}</h4>
              </div>
            </div>
            <SliderWithTooltip
              dots
              min={0}
              max={9}
              step={1}
              defaultValue={risk}
              onChange={changeRisk}
            />
          </div>
        </div>
      </div>
      <div className="row mt-3">
        <div className="col-md-12">
          <Chart data={data} />
        </div>
      </div>
    </div>
  );
};

export default ChartContainer;
