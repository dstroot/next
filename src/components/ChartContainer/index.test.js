import React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';
import ChartContainer from '../ChartContainer';

function createNodeMock() {
  const doc = document.implementation.createHTMLDocument();
  return { parentElement: doc.body };
}

describe('chart', () => {
  // mount the component
  let mountedComponent;
  const getMountedComponent = () => {
    if (!mountedComponent) {
      mountedComponent = mount(<ChartContainer />);
    }
    return mountedComponent;
  };

  beforeEach(() => {
    mountedComponent = undefined;
  });

  it('it should render', () => {
    let tree = create(<ChartContainer />, { createNodeMock });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('it should say "Design Your Future"', () => {
    const h1 = getMountedComponent()
      .find('h1')
      .first();

    expect(h1.text()).toContain('Design Your Future');
  });
});

// describe('futureValue', () => {
//   let testData = [
//     {
//       test: 1,
//       rate: 0.05,
//       nper: 12,
//       pmt: 1000,
//       pv: 0,
//       type: 0,
//       expected: 12.0,
//     },
//   ];

//   it('it should render', () => {
//     // for each case
//     let result = futureValue(
//       test[i].rate,
//       test[i].nper,
//       test[i].pmt,
//       test[i].pv,
//       test[i].type
//     );
//     expect(result === test[i].expected);
//   });
// });

// Testing information
// ---------------------------------------------------
// Jest - Expect: https://jestjs.io/docs/en/expect
// https://jestjs.io/docs/en/expect.html#content
// Enzyme cheatsheet - https://devhints.io/enzyme
// Jest cheatsheet - https://github.com/sapegin/jest-cheat-sheet

// https://medium.com/@stipsan/testing-with-jest-15-awesome-tips-and-tricks-42150ec4c262
// https://www.youtube.com/watch?time_continue=459&v=XUlGzJLZe2Q
// https://medium.freecodecamp.org/the-right-way-to-test-react-components-548a4736ab22
// https://blog.pragmatists.com/click-your-element-how-to-make-enzyme-tests-cleaner-3e7ad21c2be0
// https://blog.bitsrc.io/how-to-test-react-components-with-jest-and-enzyme-in-depth-145fcd06b90
// https://medium.com/front-end-hacking/tested-react-lets-build-a-data-table-a76aa100d23f
// https://hackernoon.com/testing-react-components-with-jest-and-enzyme-41d592c174f
// https://www.youtube.com/watch?v=59Ndb3YkLKA&feature=youtu.be

// Component Contracts
// --------------------------------------------------
// In order to test a component you must first understand what its
// Contract is. Understanding a component’s contract is the most important
// part of testing a React component. A contract defines the expected behavior
// of your component and what assumptions are reasonable to have about its
// usage. Without a clear contract, your component may be hard to understand.
// Writing tests is a great way to formally define your component’s contract.

// Every React component has at least one thing that contributes to the
// definition of its contract:

// - What it renders (which may be nothing)
//
// Additionally, most component contracts are affected by these things as well:
//
// - The props the component receives
// - The state the component holds (if any)
// - What the component does when the user interacts with it
//   (via clicking, dragging, keyboard input, etc)
