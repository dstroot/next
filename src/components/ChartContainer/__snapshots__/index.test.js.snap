// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chart it should render 1`] = `
<div
  className="container mt-5 mb-4"
>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <svg
        className="d-block w-100"
        height={300}
      >
        saving3.svg
      </svg>
    </div>
    <div
      className="col-md-6 text-center"
    >
      <h1
        style={
          Object {
            "fontSize": "calc(36px + 1vw)",
            "fontWeight": "300",
            "marginBottom": "0",
            "marginTop": "30",
          }
        }
      >
        Design Your Future
      </h1>
      <p
        className="lead text-muted mt-0"
      >
        See how fast your money can grow!
      </p>
      <div
        className="mt-4"
      >
        <div
          className="d-flex justify-content-between"
        >
          <div
            className="text-muted"
          >
            Initial Deposit
          </div>
          <div>
            <h4
              className="text-primary"
            >
              $50,000
            </h4>
          </div>
        </div>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "60.60606060606061%",
              }
            }
          />
          <div
            className="rc-slider-step"
          />
          <div
            aria-disabled={false}
            aria-valuemax={99}
            aria-valuemin={0}
            aria-valuenow={60}
            className="rc-slider-handle"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="slider"
            style={
              Object {
                "left": "60.60606060606061%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
      <div
        className="mt-4"
      >
        <div
          className="d-flex justify-content-between"
        >
          <div
            className="text-muted"
          >
            Monthly Deposit
          </div>
          <div>
            <h4
              className="text-primary"
            >
              $1,000
            </h4>
          </div>
        </div>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "53.535353535353536%",
              }
            }
          />
          <div
            className="rc-slider-step"
          />
          <div
            aria-disabled={false}
            aria-valuemax={99}
            aria-valuemin={0}
            aria-valuenow={53}
            className="rc-slider-handle"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="slider"
            style={
              Object {
                "left": "53.535353535353536%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
      <div
        className="mt-4"
      >
        <div
          className="d-flex justify-content-between"
        >
          <div
            className="text-muted"
          >
            Risk Level
          </div>
          <div>
            <h4
              className="text-primary"
            >
              Growth
            </h4>
          </div>
        </div>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "44.44444444444444%",
              }
            }
          />
          <div
            className="rc-slider-step"
          >
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "0%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "11.11111111111111%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "22.22222222222222%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "33.33333333333333%",
                }
              }
            />
            <span
              className="rc-slider-dot rc-slider-dot-active"
              style={
                Object {
                  "left": "44.44444444444444%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "55.55555555555556%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "66.66666666666666%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "77.77777777777779%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "88.88888888888889%",
                }
              }
            />
            <span
              className="rc-slider-dot"
              style={
                Object {
                  "left": "100%",
                }
              }
            />
          </div>
          <div
            aria-disabled={false}
            aria-valuemax={9}
            aria-valuemin={0}
            aria-valuenow={4}
            className="rc-slider-handle"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            role="slider"
            style={
              Object {
                "left": "44.44444444444444%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-3"
  >
    <div
      className="col-md-12"
    >
      <div
        className="recharts-responsive-container"
        style={
          Object {
            "height": 400,
            "maxHeight": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "none",
              "height": 0,
              "position": "absolute",
              "visibility": "hidden",
              "width": 0,
            }
          }
        />
      </div>
      <div
        className="d-flex justify-content-between mb-3 text-muted"
      >
        <div
          className="p-2 font-weight-bold ml-5"
        >
          Today
        </div>
        <div
          className="p-2 font-weight-bold"
        >
          30 Years
        </div>
      </div>
    </div>
  </div>
</div>
`;
